GH-125 event_webhook module for webhook sign verification

abhupadh opened the pull request:
## Description

- adds new module `events_webhook` which has the `I/O Events` webhook signature verification method

## Related Issues

* GH-125

## Motivation and Context

Helps I/O Events Customers to setup their webhook verification using this sdk feature. 

## How Has This Been Tested?

Has the test `EventVerifierTest` which tests the api for verifying signature.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the [code style](CODE_STYLE.md) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

francoisledroff commented:
@abhupadh thanks for the PR.
Let's make this simpler, let's get rid of the retrofit and caffeine dependencies.
we could implement in-memory cache with just a hash map.
for http related activites, please use either the plain java sdk (https://docs.oracle.com/en/java/javase/11/docs/api/java.net.http/java/net/http/package-summary.html) or eventually openFeign as it is already a dependency of this sdk.

abhupadh commented:
okay, will make the changes

francoisledroff reviewed (changes requested):
Could we get rid of the caffeine dependency ?
and use a Hashmap instead if you strongly feel there is a need for a cache ?
could you also push your change your PR an push to a main remote https://github.com/adobe/aio-lib-java repo branch so the integration test can be ran ?

abhupadh commented:
> Could we get rid of the caffeine dependency ? and use a Hashmap instead if you strongly feel there is a need for a cache ? could you also push your change your PR an push to a main remote https://github.com/adobe/aio-lib-java repo branch so the integration test can be ran ?

Yes, I removed the caffeine dependency and used a hashmap instead. Also, removed the retrofit with open feign. Check my latest commits. I have these tests failing which I need to fix. So, for that I think you mentioned to push this change to main branch?

francoisledroff commented:
> > Could we get rid of the caffeine dependency ? and use a Hashmap instead if you strongly feel there is a need for a cache ? could you also push your change your PR an push to a main remote https://github.com/adobe/aio-lib-java repo branch so the integration test can be ran ?
> 
> Yes, I removed the caffeine dependency and used a hashmap instead. Also, removed the retrofit with open feign. Check my latest commits. I have these tests failing which I need to fix. So, for that I think you mentioned to push this change to main branch?

I guess you forgot to push, I still see caffeine Utils and dependencies ...
yes to be able to run the integration test you will need to have the PR branch pushed to this origin repo

francoisledroff commented on pom.xml:
```diff
@@ -104,6 +104,17 @@
     <jackson.version>2.12.3</jackson.version>
     <jjwt.version>0.11.2</jjwt.version>
+    <adapter-rxjava2.version>2.7.2</adapter-rxjava2.version>
+    <retrofit.version>2.5.0</retrofit.version>
```
please let's get rid of retrofit

francoisledroff commented on pom.xml:
```diff
@@ -104,6 +104,17 @@
+    <converter-jackson.version>2.7.2</converter-jackson.version>
+    <converter-scalars.version>2.7.2</converter-scalars.version>
+    <caffeine.version>2.6.2</caffeine.version>
+    <guava.version>29.0-jre</guava.version>
```
let's get rid of rxjava, spring, h2, guava, and caffeine dependencies as well

francoisledroff commented:
I submitted a PR against your branch showing you the way
please continue, polishing it and fixing the unit test

abhupadh commented on events_webhook/src/main/java/com/adobe/aio/event/webhook/service/EventVerifier.java:
```diff
@@ -78,56 +80,46 @@ public boolean authenticateEvent(String message, String clientId,
+        Signature sig = Signature.getInstance("SHA256withRSA");
+        sig.initVerify(getPublic(fetchPemEncodedPublicKey(publicKeyPaths[i])));
+        sig.update(data);
+        boolean isSignValid = sig.verify(sign);
```
this way is required as we want to check for the next signature if the first signature is invalid

francoisledroff replied:
I would rather see a simpler and more re-usable public static method such as the one documented here https://developer.adobe.com/events/docs/guides/#security-considerations
with a more straightforward and predictable behavior (that does not assume any cache and does one header verification only ) something like
`verifySignature(String message, String publicKeyPath, String signature)`

abhupadh commented on events_webhook/src/main/java/com/adobe/aio/event/webhook/service/EventVerifier.java:
```diff
@@ -41,19 +39,24 @@ public class EventVerifier {
 
```
`@Service` / component annotation is required so that it can be auto injected in `eg-auditor`'s -> `EventServiceImpl` as here in this PR

shikhartanwar replied:
If this is the only reason that we need a @service/ @component annotation, I don't see it as a compelling reason to include the Spring framework dependency. 
Why can't we simply create an instance of `EventVerifier` in eg-auditor?

abhupadh commented:
@francoisledroff : the integration tests are failing for the `ims` module. not sure why

francoisledroff commented:
please use my branch I submitted a PR against your branch https://github.com/abhupadh/aio-lib-java/pull/1 
showing you the way, please continue, polishing it and fixing the unit test ...
Let's not have any spring dependency, nor cache, 
IMHO it's up to the customer to decide how to implement cache and whether or not  to use spring ....

abhupadh commented:
This is not only used by customers but our own eg-auditor as well.

francoisledroff commented on events_webhook/src/main/java/com/adobe/aio/event/webhook/cache/CacheService.java:
```diff
@@ -0,0 +1,27 @@
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public interface CacheService {
```
I would really like to get rid of this cache.
our customer can easily use any cache service/framework they want.
this PR should only be about showing how to validate the message headers' signature and no more

francoisledroff commented on pom.xml:
```diff
@@ -212,6 +212,16 @@
         <artifactId>feign-form</artifactId>
         <version>${feign-form.version}</version>
       </dependency>
+      <dependency>
```
let's remove these unwanted `spring` and `findbugs` dependencies

francoisledroff commented:
closing this as it was replaced by https://github.com/adobe/aio-lib-java/pull/144
